home-service/src/main/java/com/softserveinc/ita/homeproject/homeservice/service/impl/ContactServiceImpl.java

```diff
@@ -0,0 +1,104 @@
+import com.softserveinc.ita.homeproject.homeservice.mapper.ServiceMapper;
+import com.softserveinc.ita.homeproject.homeservice.service.ContactService;
+import com.softserveinc.ita.homeproject.homeservice.service.UserService;
+import lombok.RequiredArgsConstructor;
+import org.springframework.data.domain.Page;
+import org.springframework.data.domain.PageRequest;
+import org.springframework.data.jpa.domain.Specification;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+@Service
+@RequiredArgsConstructor
+public class ContactServiceImpl implements ContactService {
+
+    private final ContactRepository contactRepository;
+
+    private final ServiceMapper mapper;
+
+    private final UserService userService;
+
+    @Override
+    public ContactDto createContact(Long userId, ContactDto createContactDto) {
+        Contact contact = mapper.convert(createContactDto, Contact.class);
+        User user = mapper.convert(userService.getUserById(userId), User.class);
+        contact.setUser(user);
+        contactRepository.save(contact);
+        return mapper.convert(contact, ContactDto.class);
+    }
+
+    @Transactional
+    @Override
+    public ContactDto updateContact(Long id, ContactDto updateContactDto) {
+        Optional<Contact> contactOptional = contactRepository.findById(id);
+        if (contactOptional.isPresent()) {
+            Contact contact = contactOptional.get();
+            ContactDto contactDto = mapper.convert(contact, ContactDto.class);
+            if (contactDto.getContactType().equals(updateContactDto.getContactType())) {
+                return checkContactType(contact, updateContactDto);
+            } else {
+                throw new IllegalArgumentException("Type of the contact doesn't match");
+            }
+        } else {
+            throw new NotFoundHomeException("User with id:" + id + " is not found");
+        }
+    }
+
+    private ContactDto checkContactType(Contact contact, ContactDto updateContactDto) {
+        if (contact.getContactType().equals(ContactType.PHONE)) {
+            return updatePhone((Phone) contact, (PhoneContactDto) updateContactDto);
+        } else if (contact.getContactType().equals(ContactType.EMAIL)) {
+            return updateEmail((Email) contact, (EmailContactDto) updateContactDto);
+        } else {
+            throw new IllegalArgumentException("Invalid type of the contact");
+        }
+    }
+
+    private ContactDto updatePhone(Phone phone, PhoneContactDto phoneContactDto) {
+        phone.setPhone(phoneContactDto.getPhone());
+        contactRepository.save(phone);
+        return mapper.convert(phone, PhoneContactDto.class);
+    }
+
+    private ContactDto updateEmail(Email email, EmailContactDto emailContactDto) {
+        email.setEmail(emailContactDto.getEmail());
+        contactRepository.save(email);
+        return mapper.convert(email, EmailContactDto.class);
+    }
+
+    @Transactional
```

Comment:
Why do we need a transaction here?

Reply:
Same as above about this transaction. Don't really need it anymore.